File: itmo/2016-17/final/auto_number.py
normal_num = set()
source_num = []

for i in range(int(input())):
    number = input()
    source_num.append(number)
    if number.find('??') == -1:
        if number in normal_num:
            print("NO")
            exit(0)
        else:
            normal_num.add(number)

for k in range(len(source_num)):
    if source_num[k].find('??') != -1:
        inserted = False
        for i in range(10):
            for j in range(10):
                if not inserted:
                    fixed = source_num[k][0:4] + str(i) + str(j) + source_num[k][6:7]
                    if fixed not in normal_num:
                        normal_num.add(fixed)
                        inserted = True
                        source_num[k] = fixed
        if not inserted:
            print("NO")
            exit(0)

print("Yes")
for num in source_num:
    print(num)

File: itmo/2016-17/final/auto_number.cpp
#include <iostream>
#include <sstream>
#include <set>
#include <string>
#include <vector>

using namespace std;

int main()
{
	set<string> numbers_fixed;
	vector<string> numbers_raw;
	int n;
	cin >> n;
	for (int i = 0; i < n; ++i)
	{
		string m;
		cin >> m;
		numbers_raw.push_back(m);
		if (m.find("??") == string::npos && !numbers_fixed.insert(m).second)
		{
			cout << "No" << endl;
			exit(0);
		}
	}

	for (string& num : numbers_raw)
	{
		if (num.find("??") != string::npos)
		{
			bool found = false;
			for (int i = 0; i < 10 && !found; ++i)
				for (int j = 0; j < 10 && !found; ++j)
				{
					ostringstream buf;
					buf << num[0] << i << j << num[3] << num[4] << num[5] << num[6];
					found = numbers_fixed.insert(buf.str()).second;
					cout << "inserted: " << buf.str() << endl;
				}
			if (!found)
			{
				cout << "No" << endl;
				exit(0);
			}
		}
	}
	
	cout << "Yes" << endl;
	for (string& num : numbers_raw)
		cout << num << endl;
}
